src/nodeDef/expressionEvaluator/node/identifier.ts

```diff
@@ -0,0 +1,89 @@
+import { NodeDef, NodeDefType } from '../../nodeDef'
+import { Survey, Surveys } from '../../../survey'
+import { NodeDefs } from '../../nodeDefs'
+import { Queue } from '../../../utils'
+import { IdentifierEvaluator } from '../../../expression/javascript/node/identifier'
+import { NodeDefExpressionContext } from '../context'
+import { IdentifierExpression } from '../../../expression'
+import { NodeDefProps } from '../..'
+
+/**
+ * Determines the actual context node def
+ * - attribute def => parent entity def
+ * - virtual entity def => source node def
+ * - entity def => entity def itself
+ */
+const findActualContextNode = (params: { survey: Survey; nodeDefContext: NodeDef<NodeDefType> }) => {
+  const { survey, nodeDefContext } = params
+  if (NodeDefs.isAttribute(nodeDefContext)) {
+    return Surveys.getNodeDefParent({ survey, nodeDef: nodeDefContext })
+  }
+  if (nodeDefContext.virtual) {
+    return Surveys.getNodeDefSource({ survey, nodeDef: nodeDefContext })
+  }
+  return nodeDefContext
+}
+
+/**
+ * Get reachable nodes, i.e. the children of the node's ancestors.
+ * NOTE: The root node is excluded, but it _should_ be an entity, so that is fine.
+ */
+const getReachableNodeDefs = (params: {
+  survey: Survey
+  nodeDefContext: NodeDef<NodeDefType>
+}): NodeDef<NodeDefType, NodeDefProps>[] => {
+  const { survey, nodeDefContext } = params
+
+  const reachableNodeDefs = []
+
+  const queue = new Queue()
+  const visitedUuids: string[] = []
+
+  const actualContextNode = findActualContextNode({ survey, nodeDefContext })
+  if (actualContextNode) queue.enqueue(actualContextNode)
+
+  while (!queue.isEmpty()) {
+    const entityDefCurrent = queue.dequeue()
+    const entityDefCurrentChildren = Surveys.getNodeDefChildren({ survey, nodeDef: entityDefCurrent })
+
+    reachableNodeDefs.push(entityDefCurrent, ...entityDefCurrentChildren)
+
+    // visit nodes inside single entities
+    queue.enqueueItems(entityDefCurrentChildren.filter(NodeDefs.isSingleEntity))
```

Comment:
instead a commend could be a function extracted

Reply:
ok, I'll change it ;)

Reply:
okey ;)